pkg/controllers/leaderworkerset_controller.go

```diff
@@ -537,51 +540,105 @@ func (r *LeaderWorkerSetReconciler) iterateReplicas(ctx context.Context, lws *le
+
+	for idx := int32(0); idx < stsReplicas; idx++ {
+		nominatedName := fmt.Sprintf("%s-%d", lws.Name, idx)
 		// It can happen that the leader pod or the worker statefulset hasn't created yet
 		// or under rebuilding, which also indicates not ready.
-		if nominatedName != sortedPods[index].Name || (!noWorkerSts && nominatedName != sortedSts[index].Name) {
-			return false
+		if nominatedName != sortedPods[idx].Name || (!noWorkerSts && nominatedName != sortedSts[idx].Name) {
+			states[idx] = replicaState{
+				ready:   false,
+				updated: false,
+			}
+			continue
 		}
 
-		podTemplateHash := revisionutils.GetRevisionKey(&sortedPods[index])
-		if !(podTemplateHash == revisionKey && podutils.PodRunningAndReady(sortedPods[index])) {
-			return false
-		}
+		leaderUpdated := revisionutils.GetRevisionKey(&sortedPods[idx]) == revisionKey
+		leaderReady := podutils.PodRunningAndReady(sortedPods[idx])
 
 		if noWorkerSts {
-			return true
+			states[idx] = replicaState{
+				ready:   leaderReady,
+				updated: leaderUpdated,
+			}
+			continue
 		}
 
-		stsTemplateHash := revisionutils.GetRevisionKey(&sortedSts[index])
-		return stsTemplateHash == revisionKey && statefulsetutils.StatefulsetReady(sortedSts[index])
+		workersUpdated := revisionutils.GetRevisionKey(&sortedSts[idx]) == revisionKey
+		workersReady := statefulsetutils.StatefulsetReady(sortedSts[idx])
+
+		states[idx] = replicaState{
+			ready:   leaderReady && workersReady,
+			updated: leaderUpdated && workersUpdated,
+		}
 	}
 
-	var skip bool
-	var continuousReadyReplicas, lwsUnreadyReplicas int32
+	return states, nil
+}
+
+func rollingUpdatePartition(states []replicaState, stsReplicas int32, rollingStep int32, currentPartition int32) int32 {
+	continuousReadyReplicas := calculateContinuousReadyReplicas(states)
 
-	for index := stsReplicas - 1; index >= 0; index-- {
-		replicaReady := processReplica(index)
-		skip = skip || !replicaReady
+	// Update up to rollingStep replicas at once.
+	var minAllowedPartition = utils.NonZeroValue(stsReplicas - continuousReadyReplicas - rollingStep)
 
-		if replicaReady && !skip {
-			continuousReadyReplicas++
+	// Replicas that are not ready or expected to be updated (since idx >= partition) given current partition.
+	// Used to determine how much the partition can be reduced.
+	var unavailable int32
+	for idx := 0; idx < int(stsReplicas-continuousReadyReplicas); idx++ {
+		if !states[idx].ready || idx >= int(currentPartition) {
+			unavailable++
+		}
+	}
+
+	// Iterate over replicas, determine if moving partition will violate the maxUnavailable (rollingStep).
+	// Note that the returned partition will always be smaller than or equal to currentPartition.
+	// When updated replicas become not ready again or scaled up replicas are not ready yet,
+	// we'll not modify the Partition field. That means Partition moves in one direction to make it simple.
+	var partition = currentPartition
+	for idx := partition - 1; idx >= minAllowedPartition; idx-- {
```

Comment:
If we agree that the to-update-partition is belong to [minAllowedPartition, currentPartition], can we iterate the [0, minAllowedPartition) to find the unavailable ones, and the to-update-partition should be `min(minAllowedPartition+unavailable, currentPartition)`.